### ISSUE TITLE
Add support for custom cluster icon in Google Map Flutter

### ISSUE THREAD

**Issue by Mohit8G** [flutter/flutter#153092]
### Use case

So Clustering in Google map is just landed but we don't have any support for custom cluster icon. There are some package on pub.dev but they are not proper solution to the problem.

### Proposal

Need some properties that can override default cluster marker.
Like this
https://developers.google.com/maps/documentation/android-sdk/utility/marker-clustering

The ClusterManager constructor creates a DefaultClusterRenderer and a NonHierarchicalDistanceBasedAlgorithm. You can change the ClusterRenderer and the Algorithm using the setAlgorithm(Algorithm<T> algorithm) and setRenderer(ClusterRenderer<T> view) methods of ClusterManager.

You can implement ClusterRenderer to customize the rendering of the clusters. DefaultClusterRenderer provides a good base to start from. By subclassing DefaultClusterRenderer, you can override the defaults.

**Comment by stuartmorgan** [comment]
It looks like this is not going to be implementable without https://issuetracker.google.com/issues/140415468 (or, alternately, https://github.com/flutter/flutter/issues/150525), since the cluster renderer protocol requires a synchronous response. Usually when we have something that requires a synchronous response we pre-generated the necessary information, send it asynchronously, and store it on the platform thread for immediate response, but there's no way to pre-generate all possible images.

Even if clients didn't want the ability to change the icon dynamically (which they well might), they will almost certainly want to add a number, and that can be arbitrary as well.

**Comment by stuartmorgan** [comment]
Please use a thumbs-up reaction on this comment specifically (in addition to the original feature request) if the following limited version of this feature would address your use case:

We could potentially implement an API that allowed pre-providing a map of `{cluster count, image bitmap}` cluster options, and have the cluster icon picked by finding the image in that map whose cluster count is the largest value that is <= the actual count. For instance, if someone provided `{1: imageA, 10: imageB, 50: imageC, 100: imageD}`, then any clusters of size 1-9 would use imageA, 10-49 would use imageB, 50-99 would use imageC, and anything larger would use imageD.

**Comment by Rishyash** [comment]
@stuartmorgan Needed one suggestion for this as I am not a native developer, what if I fork this package and add custom cluster logic for my usecase on native side code will this reflect on the flutter side map view?

**Comment by kaj777** [comment]
Can we expect the feature anytime soon? Custom cluster icon is a must for us.

**Comment by PauloAlves311** [comment]
I'm really happy that there is now official support for clustering, but the inability to alter the clusterOptions leaves it unusable in my opinion. Specially with custom map styles, the variety of colors makes it look really out of place.  But let's be honest, green, red, blue, brown... Even without a custom map style, these colors look awful when put together, I can't believe any dev that cares about design would use these clusters as it is.

The clusterIcons also look too big in my personal opinion, so it would be nice to be able to adjust their scaling.

I am really not familiar with the limitations to implement this, but ideally there would be an example of the default renderer which would allow us to tweak all the details. As I can in js or with Fluster, but until then I really don't think I can use the official clustering.

**Comment by lucavenir** [comment]
I'm not sure what's the blocker, but I used to be able to do this with custom packages such as [this one](https://pub.dev/packages/google_maps_cluster_manager).

Now this package is essentially incompatible with the new APIs, but in return I can't customize my pins anymore.

Why did this feature land in the first place?

**Comment by FeelHippo** [comment]
I agree that clustering without the possibility of customizing the appearance is not very useful. However, I spent a bit of time looking into a possible solution, and it's definitely something I cannot fix. 

I understand that the ClusterManager constructor [here](https://github.com/flutter/packages/blob/main/packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java) should probably[ extend setAlgorithm and setRenderer](https://developers.google.com/maps/documentation/android-sdk/utility/marker-clustering#custom), and then trickle that down all the way back to the [interface](https://github.com/flutter/packages/blob/ca8cb667042bdf485601be2a51eccde4692d54b3/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/cluster_manager.dart#L20)?

Not really my department, but I might give it a go. Or has @stuartmorgan been working on this?

**Comment by Shajk0** [comment]
When I read that flutter google maps offers its own clustering option, I immediately tried to implement it in my project, but when I realized that the cluster icons are not changeable and the numbers are fixed, it makes the whole thing completely unusable.
I would be happy if you create the possibility to make the cluster icons freely switchable just like the marker icons and also provide the possibility that the numbers are not displayed in these 10,50,100 groups but for example in the exact amount of the cluster

**Comment by Homerwatch** [comment]
Also facing this blocker, kind of unacceptable for such a big addition to be doing the basics. Any news on if and when this can be added?

**Comment by Homerwatch** [comment]
any news on this? This is practically a beta library we have no ability to modify the icon of cluster

**Comment by stefanvinteler** [comment]
Any news?

**Comment by mangeria** [comment]
Is it even possible to change the cluster default colors? Just implemented this on my app but the lack of customization options means I might have to look for another solutions.

**Comment by Mohit8G** [comment]
> Is it even possible to change the cluster default colors? Just implemented this on my app but the lack of customization options means I might have to look for another solutions.

For now no customization can be made.

**Comment by mangeria** [comment]
> > Is it even possible to change the cluster default colors? Just implemented this on my app but the lack of customization options means I might have to look for another solutions.
> 
> For now no customization can be made.

It would be beneficial to at least have things like:

1. Cluster Nesting. Whichever is better implementation from following two

- Cluster Zoom. Markers will cluster when the map is zoomed out beyond certain level. Currently I have to be zoomed all the way to world map for cluster market to appear in countries, markers inside cities are not clustered at all.
- Cluster Grouping. Controls how close markers need to be to form a cluster. Unit could be radius?

2. Custom Cluster Colors
3. Custom Cluster Text

I'll keep following this issue. Thanks for consideration.

**Comment by Dzivo** [comment]
Omg you are slow this was needed 5 years ago XD

**Comment by servefast-cto** [comment]
Cant wait for this my app looks ugly but clustering works really well

**Comment by niemsie** [comment]
Still no updates? It's been over half a year and no new features for the ClusterManager?

Besides marker customisation it would also be nice to be able to set a max zoom level at which clustering is disabled (e.g. if there are some pins that are just too close to each other).

**Comment by servefast-cto** [comment]
I guess maps just arent that important to them even though i haven't build an app without one

**Comment by Homerwatch** [comment]
Still can't wrap my head around the fact that almighty Google can't even support their own language. And is not even their language, maps is their own product too. This is so frustrating guys

**Comment by Dzivo** [comment]
@Homerwatch  We would accept any temp solution until they come up with a better one. What i don't understand is they have such high standards for employment who approved cluster feature without even trying it. i have removed it and used a third party library thats not as good as this but at least i can customize it

**Comment by ivanpolish123** [comment]
Whats the status of this ?

**Comment by andynvt** [comment]
Still waiting for this. Or is it possible to set the cluster color?

**Comment by Dzivo** [comment]
Haha your comment on new feature "Nothing broken request for a new capability" it never worked in the first place

**Comment by Dzivo** [comment]
How is this a new feature ?

**Comment by andynvt** [comment]
up please, it shows the default cluster marker only

**Comment by dimonkomaron** [comment]
> [#150525](https://github.com/flutter/flutter/issues/150525)

@stuartmorgan-g Looks like it was implemented in newest Flutter version and customised clusters can be implemented, or are there any other obstacles?

**Comment by stuartmorgan-g** [comment]
Plugins code cannot rely on thread merging while there is an application-level opt-out for it, because they have no way of knowing what mode they will be running in.

**Comment by matthew-niemann** [comment]
We will get GTA 6 before custom cluster markers.

**Comment by Dzivo** [comment]
Can we start with at lease color change ? Then move to custom elements later ?

**Comment by Dzivo** [comment]
No wonder flutter apps look bad when we cant get this simple feature

**Comment by Coinners** [comment]
are there still plants on implementation? at least for a limited workaround?

**Comment by zombie6888** [comment]
> https://pub.dev/packages/google_maps_cluster_manager
> 
> Already supports adding a custom icon to a cluster. Maybe we can take some inspiration from there?

I think that's a bad idea. I tried this package. They make cluster from markers, there is no cluster appearance animation and entire code becomes more complicated. I think, it should be created on native side as google supports it
